api/apps/geoprocessing/src/modules/scenario-planning-units-inclusion/scenario-planning-units-inclusion-processor.ts

```diff
@@ -1,12 +1,108 @@
 import { JobInput } from '@marxan-jobs/planning-unit-geometry';
+import {
+  LockStatus,
+  ScenariosPlanningUnitGeoEntity,
+} from '@marxan/scenarios-planning-unit';
 
 @Injectable()
 export class ScenarioPlanningUnitsInclusionProcessor
   implements WorkerProcessor<JobInput, true> {
-  async process(_job: Job<JobInput, true>): Promise<true> {
+  constructor(
+    @InjectRepository(ScenariosPlanningUnitGeoEntity)
+    private readonly scenarioPlanningUnitsRepo: Repository<ScenariosPlanningUnitGeoEntity>,
+  ) {}
+
+  async process(job: Job<JobInput, true>): Promise<true> {
+    const scenarioId = job.data.scenarioId;
+    const includeGeo = job.data.include?.geo;
+    const excludeGeo = job.data.exclude?.geo;
+
+    const geometriesIdsToInclude: string[] = [];
+    const geometriesIdsToExclude: string[] = [];
+
+    if (includeGeo) {
+      const targetGeometries = flatten(
+        includeGeo.map((collection) => collection.features),
+      ).map((feature) => feature.geometry);
+      geometriesIdsToInclude.push(
+        ...(
+          await this.getIntersectedGeometriesFor(scenarioId, targetGeometries)
+        ).map(({ pu_geom_id }) => pu_geom_id),
+      );
+    }
+
+    if (excludeGeo) {
+      const targetGeometries = flatten(
+        excludeGeo.map((collection) => collection.features),
+      ).map((feature) => feature.geometry);
+      geometriesIdsToExclude.push(
+        ...(
+          await this.getIntersectedGeometriesFor(scenarioId, targetGeometries)
+        ).map(({ pu_geom_id }) => pu_geom_id),
+      );
+    }
+
+    await this.scenarioPlanningUnitsRepo.update(
+      {
+        scenarioId,
+      },
+      {
+        lockStatus: LockStatus.Unstated,
+      },
+    );
+
+    await this.scenarioPlanningUnitsRepo.update(
+      {
+        scenarioId,
+        puGeometryId: In(geometriesIdsToInclude),
+      },
+      {
+        lockStatus: LockStatus.LockedIn,
+      },
+    );
+
+    await this.scenarioPlanningUnitsRepo.update(
+      {
+        scenarioId,
+        puGeometryId: In(geometriesIdsToExclude),
+      },
+      {
+        lockStatus: LockStatus.LockedOut,
+      },
+    );
     return true;
   }
+
+  private async getIntersectedGeometriesFor(
+    scenarioId: string,
+    geometries: (Polygon | MultiPolygon)[],
+  ): Promise<Array<{ pu_geom_id: string }>> {
+    const queryBuilder = this.scenarioPlanningUnitsRepo
+      .createQueryBuilder(`spd`)
+      .select(['spd.scenario_id', 'spd.pu_geom_id', 'spd.id'])
+      .leftJoin(`planning_units_geom`, `pug`, `pug.id = spd.pu_geom_id`)
+      .where(`spd.scenario_id = :scenarioId`, { scenarioId })
+      .andWhere(
+        new Brackets((qb) =>
+          geometries.reduce<WhereExpression>(
+            (queryBuilder, geom, index) =>
+              qb.orWhere(
+                `st_intersects(ST_GeomFromGeoJSON(:geom${index}),pug.the_geom)`,
+                { [`geom${index}`]: JSON.stringify(geom) },
+              ),
```

Comment:
I wonder what is better, this (`OR...OR...OR`) vs `st_intersects(st_union(all geometries), pug.the_geom`
sometimes dbs don't like `OR`s

Reply:
@Dyostiq not sure if it would be the same - it should return all intersected nodes; if we would union them beforehand, wouldn't it spoil the desired results?

Reply:
(A∩C)∪(B∩C)=(A∪B)∩C
unless it doesn't work for geo/geojsons 😅

Reply:
You are right, not sure tho which is better 🤷 Would shoot that union would do...

Reply:
I would go for intersection of union... @aagm your take here?

Reply:
adjusted in https://github.com/Vizzuality/marxan-cloud/pull/258/commits/40c4026b85c2d0779bbba9cfd019da1d6ff35b80